sources/lib/run-time/llvm-exceptions.c

```diff
@@ -148,14 +148,185 @@ void RemoveDylanExceptionHandlers(void)
+    }
+    else if (code[1] & FE_OVERFLOW) {
+      handler = (uintptr_t) &dylan_float_overflow_error;
+    }
+    else if (code[1] & FE_UNDERFLOW) {
+      handler = (uintptr_t) &dylan_float_underflow_error;
+    }
+    else {
+      fprintf(stderr, "Unhandled EXC_ARITHMETIC: code=%lld/%lld\n", 
+              code[0], code[1]);
+      abort();
+    }
+    break;
+
+  default:
+    fprintf(stderr, "Unhandled EXC_ARITHMETIC: code=%lld/%lld\n", 
+            code[0], code[1]);
+    abort();
+  }
+
+#if defined OPEN_DYLAN_ARCH_X86
+  x86_thread_state32_t *ts = (x86_thread_state32_t *) old_state;
+  uintptr_t *esp = (uintptr_t *) (ts->__esp & -STACK_ALIGNMENT);
+  *--esp = ts->__eip;
+  ts->__esp = (uintptr_t) esp;
+  ts->__eip = handler;
+#elif defined OPEN_DYLAN_ARCH_X86_64
+  x86_thread_state64_t *ts = (x86_thread_state64_t *) old_state;
+  uintptr_t *rsp = (uintptr_t *) (ts->__rsp & -STACK_ALIGNMENT);
+  *--rsp = ts->__rip;
+  ts->__rsp = (uintptr_t) rsp;
+  ts->__rip = handler;
+#else
+#error No definition for catch_mach_exception_raise_state_identity provided
+#endif
+  *new_stateCnt = old_stateCnt;
+  memcpy(new_state, old_state, old_stateCnt * sizeof(natural_t));
+
+  return KERN_SUCCESS;
+}
+
 void EstablishDylanExceptionHandlers(void)
 {
-  // FIXME
+  if (exception_port == MACH_PORT_NULL) {
+    // Need a port we can receive exceptions on
+    kern_return_t rc
+      = mach_port_allocate(mach_task_self(), MACH_PORT_RIGHT_RECEIVE,
+                           &exception_port);
+    if (rc != KERN_SUCCESS) {
+      mach_error("mach_port_allocate send", rc);
+      abort();
+    }
+
+    // Need to be able to send on it too
+    rc = mach_port_insert_right(mach_task_self(), exception_port, exception_port,
+                                MACH_MSG_TYPE_MAKE_SEND);
+    if (rc != KERN_SUCCESS) {
+      mach_error("mach_port_insert_right", rc);
+      abort();
+    }
+
+    // Spawn a thread to serve exception requests
+    pthread_attr_t attr;
+    pthread_attr_init(&attr);
+    pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+    pthread_t thread;
+    int prc = pthread_create(&thread, &attr, catcher, NULL);
+    pthread_attr_destroy(&attr);
+    if (prc != 0) {
+      fprintf(stderr, "%s: pthread_create returned %d\n",
+              __func__, prc);
+      abort();
+    }
+  }
+
+  // Set this thread's exception port
+  kern_return_t rc
+    = thread_set_exception_ports(mach_thread_self(),
+                                 EXC_MASK_ARITHMETIC,
+                                 exception_port,
+                                 EXCEPTION_STATE_IDENTITY|MACH_EXCEPTION_CODES,
+                                 THREAD_STATE_FLAVOR);
+  if (rc != KERN_SUCCESS) {
+    mach_error("thread_set_exception_ports", rc);
+    abort();
+  }
+
+  primitive_reset_float_environment();
 }
 
 void RemoveDylanExceptionHandlers(void)
 {
-  // FIXME
+  // Nothing to do (thread is about to exit)
```

Comment:
That depends... calling into Dylan from the main thread, or from a thread started using make(<thread>), should be just fine. Calling into Dylan from a foreign thread would require providing a trampoline interface which would take care of this sort of thing as well as cooperation with the GC. If there is a need for such an interface we can reexamine what this function does.